Refine output where we have type data

trueadm opened the pull request:
Release notes: none

In some places we have type information but don't fully use the benefits it brings and instead go down a non-optimal path. This PR addresses two fairly simple cases where this occurs:

- when we have a method call on an abstract value that has a primitive type, where the name of the method matches that of one of the methods on the primitive prototype, then use that method.
- when we have a type of "array", use all the logic for unknown abstract arrays rather than generate an abstract value with a type of array (which has no optimized paths, unlike unknown arrays).
- adds temporals for `toString` and other built-ins because of existing logic around joining them and the fact that they can throw if called on `undefined`, it makes them unsafe. Previous we rarely went down these code paths to create template abstracts, because we guarded in pure scope for abstract base values in `CallExpression`.

calebmer commented on src/evaluators/CallExpression.js:
```diff
@@ -79,6 +100,21 @@ function evaluateReference(
+    // if we are referencing a prototype method, then it's safe to access, even
+    // on an abstract value as the value is immutable and can't have a property
+    // that matches the prototype method (unless the prototype was modified).
+    // In pure scope, we assume the global prototype of built-ins has not been altered.
```
Do we want a `realm.isInPureScope()` check for this?

trueadm commented:
This PR is still good to merge, we just can't apply the type data changes to our internal bundle that add types for arrays, because we need to have materialization as a feature added (without havocing/leaking anything in the process).

cblappert commented on src/evaluators/CallExpression.js:
```diff
@@ -79,6 +100,21 @@ function evaluateReference(
+    // if we are referencing a prototype method, then it's safe to access, even
+    // on an abstract value as the value is immutable and can't have a property
+    // that matches the prototype method (unless the prototype was modified).
+    // In pure scope, we assume the global prototype of built-ins has not been altered.
```
I was not aware of this, can we document this somewhere in the code or the wiki for pure evaluation mode?

trueadm replied:
It should be documented on https://github.com/facebook/prepack/wiki/Compiler-assumptions

cblappert commented on src/evaluators/CallExpression.js:
```diff
@@ -79,6 +100,21 @@ function evaluateReference(
+    // In pure scope, we assume the global prototype of built-ins has not been altered.
+    let possiblePrimitiveProto = getPrimitiveProtoFromAbstractValueType(realm, base);
+    if (possiblePrimitiveProto !== undefined && typeof referencedName === "string") {
+      let possibleProtoDesc = possiblePrimitiveProto.$GetOwnProperty(referencedName);
```
Do we also make the assumption in pure mode that we don't have properties on the abstract value with the same name as global prototype properties?

trueadm replied:
They are primitive values, they can't have properties (that I'm aware of) on `string`, `number`, `boolean` and `symbol`. You can `[]` access characters at an index, but then that wouldn't be on the prototype.

hermanventer commented on src/intrinsics/ecma262/ArrayPrototype.js:
```diff
@@ -992,6 +992,8 @@ export default function(realm: Realm, obj: ObjectValue): void {
         args.push(thisArg);
       }
+      // We need to materialize here, but there's no way of doing it without
+      // havocing, which brings about even more issues. :/
```
This comment creates FUD rather than enlightenment. Can you reference an issue that spells out the issue in detail, along with examples?

trueadm replied:
I can make a repro as a separate issue.

NTillmann replied:
And then please rewrite this comment and mention the issue number. As it stands right now, it's an odd comment to have in the code base.

hermanventer commented on src/intrinsics/ecma262/BooleanPrototype.js:
```diff
@@ -12,20 +12,23 @@
     const target = context instanceof ObjectValue ? context.$BooleanData : context;
     if (target instanceof AbstractValue && target.getType() === BooleanValue) {
-      return AbstractValue.createFromTemplate(realm, tsTemplate, StringValue, [target], tsTemplateSrc);
+      return AbstractValue.createTemporalFromBuildFunction(
```
This makes no sense to me. We know that target is a primitive **boolean** (not undefined and not null) and we know that this version of toString has no side effects. Calling it via the time line is just not right.

There is a bit of an issue with the existing code: All we really know at this point is that there is an expression that ends up calling this toString method on the original value of String.prototype.toString. We don't actually know that (A).toString() will end up calling this method at runtime. So, to be on the safe side we should guard this case with the additional constraint that the current value of String.prototype.toString is in fact this method.

trueadm replied:
I was probably a bit eager here and didn't need to change this one.

hermanventer replied:
None of the String.prototype.methods need to become temporal.

hermanventer commented on src/values/AbstractObjectValue.js:
```diff
@@ -515,10 +516,18 @@ export default class AbstractObjectValue extends AbstractValue {
             propertyGetter = propertyShape.getGetter();
           }
         }
+        // Create an unknown array temporal
```
Are there cases where we do not want createTemporalFromBuildFunction to return a value that is the closest match to its type argument? If not, it seems to me that ArrayValue.createTemporalWithWidenedNumericProperty should be invoked via createTemporalFromBuildFunction and never directly.

trueadm replied:
Maybe we can at a future point, I don't feel this PR should change all those call sites though, there are quite a fair few of them.

hermanventer commented on src/evaluators/CallExpression.js:
```diff
@@ -106,10 +106,11 @@ function evaluateReference(
+    // global code has finished.
+    let prototypeIfPrimitive = getPrimitivePrototypeFromType(realm, base);
+    if (prototypeIfPrimitive !== undefined && typeof referencedName === "string") {
+      let possibleProtoDesc = prototypeIfPrimitive._SafeGetDataPropertyValue(referencedName);
```
_SafeGetDataPropertyValue returns a Value, not a descriptor.

trueadm commented:
@hermanventer I can revert the changes for the values to temporals and keep them as templates and try and come up with a repro case that I was seeing on our internal bundle?

trueadm commented:
Test case that fails with non-temporal `toString` and `split`:

Also related to our chat on workplace:
@NTillmann @hermanventer @sebmarkbage 

```js

function fn(cond, cond2, a, b, c) {
  var x = global.__abstract ? __abstract("number", "a") : a;
  var y;
  var z;
  var obj = Object.assign({}, b, {x})

  if (cond) {
    y = obj.x.toString() + c;
    z = y.split("");
  } else {
    if (cond2) {
      return null;
    }
    y = obj.x.toString() + " " + c;
    z = y.split("");
  }
  return z;
}

inspect = function() {
  return fn(false, false, 2, {}, 5);
};

this.__optimize && __optimize(fn);
```

trueadm commented:
A better case that fails even with `inlineExpressions` enabled, this appears to be a clear bug.

Input:
```js
function fn2(cond, items) {
  if (cond) {
    return items.length;
  }
  return 0;
}

function fn(cond, x, y) {
  var items = Array.from(x);
  var items2 = Array.from(y);

  var len = fn2(cond, items);
  var len2 = fn2(cond, items2);

  if (len > 0) {
    return len.toString();
  }
  if (len2 > 0) {
    return len.toString();
  }
  return "Should hit this!";
}

this.__optimize && __optimize(fn);

inspect = function() {
  return fn(false, [], []);
}
```
Output:

```js
  var _4 = function (cond, x, y) {
    var _O = _$6;

    var _$0 = _O(x);

    var _$1 = _O(y);

    if (cond) {
      var _$2 = _$0.length;
    }

    if (cond) {
      var _$3 = _$1.length;
    }

    var _7 = _$2.toString();  //  <---$2 might not be declared....

    return (cond ? _$2 > 0 : false) ? _7 : (cond ? _$3 > 0 : false) ? _7 : "Should hit this!";
  };
```

hermanventer commented:
Please ensure that the test suite succeeds. Just change to failing test to expect failure.

trueadm commented:
I can’t land this as it fails our internal bundle test with the same problems outlined in work chat.

trueadm commented:
Some progress, as per our discussion in https://github.com/facebook/prepack/issues/2327. I've made the following changes:

- Made `String.prototype.toString` output `('' + A)` as the template, rather than `(A).toString()` like before. Doing so makes all the tests around these issues pass and actually has the nice effect of being slightly more performant than using `toString()` in V8 and JSC.
- Found a bug in `String.prototype.split` where the `resultType` was incorrectly set to `StringValue` when it should have been `ArrayValue`.
- Updated `AbstractValue.createFromTemplate` to apply the same hashing logic when the `resultType` is that of an `ObjectValue` as `createFromType` already did.
- Fixed a bug where `createFromType` only checked if the `resultType` was that of `ObjectValue`, but this misses cases where the `resultType` might be a subclass of `ObjectValue`.

NTillmann commented on src/intrinsics/ecma262/NumberPrototype.js:
```diff
@@ -144,7 +144,7 @@ export default function(realm: Realm, obj: ObjectValue): void {
   });
 
-  const tsTemplateSrc = "(A).toString()";
+  const tsTemplateSrc = "('' + A)";
```
Annoying to have this so many times in our code base. Time to consolidate?

trueadm replied:
There's only three cases of them and they're backed by the Flow enum. Extracting it out into a single place is a good idea though, but it also makes all the other template strings inconsistent – they should all then be imported from a central location to be consistent. Something for a follow up I guess.

NTillmann commented on src/values/AbstractValue.js:
```diff
@@ -729,6 +730,7 @@ export default class AbstractValue extends Value {
     let labels = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
     invariant(labels.length >= operands.length);
     let operationDescriptor = createOperationDescriptor("ABSTRACT_FROM_TEMPLATE", { template });
+    if (resultType === ObjectValue || resultType === ArrayValue) hash = ++realm.objectCount;
```
What about `FunctionValue`? Do you want `Value.isTypeCompatibleWith(resultType, ObjectValue)` that's already mentioned above?

In itself, an object result doesn't mean that the function isn't pure, this is just being conservative. Leave behind a comment that mentions this (future optimization opportunity).

NTillmann commented on src/values/AbstractValue.js:
```diff
@@ -744,7 +746,7 @@ export default class AbstractValue extends Value {
     let Constructor = Value.isTypeCompatibleWith(resultType, ObjectValue) ? AbstractObjectValue : AbstractValue;
     let [hash, args] = hashCall(resultType.name + (kind || ""), ...(operands || []));
-    if (resultType === ObjectValue) hash = ++realm.objectCount;
+    if (resultType === ObjectValue || resultType === ArrayValue) hash = ++realm.objectCount;
```
Same as above.

NTillmann commented on src/evaluators/CallExpression.js:
```diff
@@ -79,6 +100,22 @@ function evaluateReference(
+    // if we are referencing a prototype method, then it's safe to access, even
+    // on an abstract value as the value is immutable and can't have a property
+    // that matches the prototype method (unless the prototype was modified).
+    // We assume the global prototype of built-ins has not been altered since
```
Mention #1233 in comment text re. unmodified global prototypes.
